oxidation/libparsec/crates/platform-async/src/wasm32/task.rs

```diff
@@ -0,0 +1,167 @@
+use std::{
+    future::Future,
+    pin::Pin,
+    sync::{Arc, Mutex},
+    task::{Context, Poll, Waker},
+};
+
+use wasm_bindgen_futures::spawn_local;
+
+/// Spawns a new asynchronous task, returning a [Task] for it.
+///
+/// ```
+/// use libparsec_platform_async::spawn;
+/// use futures::ready;
+///
+/// # futures_lite::future::block_on(async {
+/// let future = ready(42);
+/// let task = spawn(future);
+/// let result = task.await
+/// assert_eq!(result, 42);
+/// # });
+/// ```
+pub fn spawn<F>(future: F) -> Task<F::Output>
+where
+    F: Future + 'static,
+    F::Output: 'static,
+{
+    let shared_state = SharedState::default();
+    let shared_state = Arc::new(Mutex::new(shared_state));
+
+    let task = Task::new(shared_state.clone());
+    let runnable = Runnable::new(future, shared_state);
+
+    spawn_local(runnable);
+
+    task
+}
+
+struct SharedState<T> {
+    canceled: bool,
+    finished: bool,
+    detached: bool,
+    value: Option<T>,
+    task_waker: Option<Waker>,
+    runnable_waker: Option<Waker>,
+}
+
+impl<T> Default for SharedState<T> {
```

Comment:
Same as #[derive(Default)]

Reply:
```
error[E0277]: the trait bound `<F as futures::Future>::Output: std::default::Default` is not satisfied
  --> oxidation/libparsec/crates/platform-async/src/native/task.rs:31:24
   |
31 |     let shared_state = SharedState::default();
   |                        ^^^^^^^^^^^^^^^^^^^^ the trait `std::default::Default` is not implemented for `<F as futures::Future>::Output`
   |
note: required because of the requirements on the impl of `std::default::Default` for `native::task::SharedState<<F as futures::Future>::Output>`
  --> oxidation/libparsec/crates/platform-async/src/native/task.rs:44:10
   |
44 | #[derive(Default)]
   |          ^^^^^^^
   = note: this error originates in the derive macro `Default` (in Nightly builds, run with -Z macro-backtrace for more info)
help: consider further restricting the associated type
   |
29 |     F::Output: Send + 'static, <F as futures::Future>::Output: std::default::Default
   |                              ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 ```

Reply:
This error is pretty strange :thinking:

Reply:
yes, it require that the output must implement `Default` but `Option::default()` to `None`

Reply:
https://doc.rust-lang.org/std/default/trait.Default.html#impl-Default-70
Default implementation for Option<T> doesn't care about what is T, so SharedState::default() is fully described, what's the joke ? :joy:

Reply:
Even if SharedState::default() doesn't know the type of T, how the custom implementation can solve this ?

Reply:
Oh ! That's why ! It can't solve the type statically
```Rust
macro_rules! default_impl {
    ($t:ty, $v:expr, $doc:tt) => {
        #[stable(feature = "rust1", since = "1.0.0")]
        #[rustc_const_unstable(feature = "const_default_impls", issue = "87864")]
        impl const Default for $t {
            #[inline]
            #[doc = $doc]
            fn default() -> $t {
                $v
            }
        }
    };
}
```